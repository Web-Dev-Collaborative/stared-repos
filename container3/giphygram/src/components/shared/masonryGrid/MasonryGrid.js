import React from 'react';
import PropTypes from 'prop-types';
import { SUPPORTED_COLUMNS_NUMS, TOTAL_NUMBER_OF_LAYOUT_COLUMNS } from '../../../services/LayoutService';

export class MasonryGrid extends React.Component {
  static propTypes = {
    renderItem: PropTypes.func.isRequired,
    columnsNum: PropTypes.oneOf(SUPPORTED_COLUMNS_NUMS).isRequired,
    items: PropTypes.arrayOf(PropTypes.shape({
      height: PropTypes.number,
      content: PropTypes.object,
    })),
  };

  static defaultProps = {
    items: [],
  };

  render() {
    const { items, columnsNum, renderItem } = this.props;

    // Array that will store the height of each column.
    // It is used to keep column heights similar.
    const columnsSizes = new Array(columnsNum).fill(0);

    // Array that will store items that are split by columns according to their heights.
    const itemsPartitions = new Array(columnsNum)
      .fill(null)
      .map(() => []);

    // Fill partitions with items.
    items.forEach((item) => {
      // Get next smallest column to place the icon into.
      const itemColumnIndex = columnsSizes.indexOf(Math.min(...columnsSizes));
      // Increase the size of the column by item height.
      columnsSizes[itemColumnIndex] += item.height;
      // Put item to the proper partition.
      itemsPartitions[itemColumnIndex].push(item.content);
    });

    // Generate columns with search items inside.
    const columns = itemsPartitions.map((partition, partitionIndex) => {
      /* eslint-disable react/no-array-index-key */

      // Calculate Bootstrap class depending on the columns num.
      const columnClass = `col-${TOTAL_NUMBER_OF_LAYOUT_COLUMNS / columnsNum}`;

      // Generate column entities.
      const columnElements = partition.map(
        (item, index) => renderItem(item, index),
      );

      // Wrap column entities into column wrapper.
      return (
        <div className={columnClass} key={partitionIndex}>
          {columnElements}
        </div>
      );
    });

    return (
      <div className="row">
        {columns}
      </div>
    );
  }
}
