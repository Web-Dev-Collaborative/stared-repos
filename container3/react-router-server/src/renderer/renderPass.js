import React from 'react';
import { renderToString, renderToStaticMarkup } from 'react-dom/server';
import AsyncRenderer from '../components/AsyncRenderer';
import removeDuplicateModules from '../utils/removeDuplicateModules';

const renderPass = (context, element, staticMarkup = false) => {
  context.callback = () => {
    if (context.finishedLoadingModules && !context.statesRenderPass) {
      context.statesRenderPass = true;
      context.renderResult = renderPass(context, element, staticMarkup);
      if (context.fetchingStates <= 0 && context.modulesLoading <= 0) {
        context.resolve({
          html: context.renderResult,
          state: context.fetchStateResults === undefined ? null : context.fetchStateResults,
          modules: context.modules === undefined ? null : removeDuplicateModules(context.modules)
        });
      }
    } else if (context.finishedLoadingModules && context.statesRenderPass || !context.hasModules) {
      context.renderResult = renderPass(context, element, staticMarkup);
      if (context.fetchingStates <= 0 && context.modulesLoading <= 0) {
        context.resolve({
          html: context.renderResult,
          state: context.fetchStateResults === undefined ? null : context.fetchStateResults,
          modules: context.modules === undefined ? null : removeDuplicateModules(context.modules)
        });
      }
    }
  };

  let component = (
    <AsyncRenderer context={context}>
      {element}
    </AsyncRenderer>
  );

  let result;
  try {
    if (staticMarkup) {
      result = renderToStaticMarkup(component);
    } else {
      result = renderToString(component);
    }
  } catch (e) {
    return context.reject(e);
  }

  if (!context.hasModules && !context.hasStates) {
    context.resolve({
      html: result,
      state: context.fetchStateResults === undefined ? null : context.fetchStateResults,
      modules: context.modules === undefined ? null : removeDuplicateModules(context.modules)
    });
  }
  return result;
};

export default renderPass;
