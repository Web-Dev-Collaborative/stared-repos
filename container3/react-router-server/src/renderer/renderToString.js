import renderPass from './renderPass';
import { clear } from '../module/cache';

export default (element) => {
  return new Promise((resolve, reject) => {
    clear();
    const context = {
      resolve, reject,
      modulesLoading: 0,
      fetchingStates: 0
    };
    renderPass(context, element);
  });
};
