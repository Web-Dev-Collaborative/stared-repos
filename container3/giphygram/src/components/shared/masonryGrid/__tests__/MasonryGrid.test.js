import React from 'react';
import renderer from 'react-test-renderer';
import { MasonryGrid } from '../MasonryGrid';

const itemsMock = [
  {
    height: 100,
    content: { id: 1, name: 'Item #1' },
  },
  {
    height: 200,
    content: { id: 2, name: 'Item #2' },
  },
  {
    height: 150,
    content: { id: 3, name: 'Item #3' },
  },
  {
    height: 250,
    content: { id: 4, name: 'Item #4' },
  },
  {
    height: 300,
    content: { id: 5, name: 'Item #5' },
  },
];

const renderItemMock = item => <div key={item.id}>{item.name}</div>;

describe('MasonryGrid', () => {
  it('should be rendered correctly with 1 column', () => {
    const tree = renderer
      .create((
        <MasonryGrid
          items={[...itemsMock]}
          renderItem={renderItemMock}
          columnsNum={1}
        />
      ))
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should be rendered correctly with 3 columns', () => {
    const tree = renderer
      .create((
        <MasonryGrid
          items={[...itemsMock]}
          renderItem={renderItemMock}
          columnsNum={3}
        />
      ))
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should be rendered correctly with 4 columns', () => {
    const tree = renderer
      .create((
        <MasonryGrid
          items={[...itemsMock]}
          renderItem={renderItemMock}
          columnsNum={4}
        />
      ))
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
