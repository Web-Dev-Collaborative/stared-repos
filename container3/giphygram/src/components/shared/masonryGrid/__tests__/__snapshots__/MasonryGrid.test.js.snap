// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MasonryGrid should be rendered correctly with 1 column 1`] = `
<div
  className="row"
>
  <div
    className="col-12"
  >
    <div>
      Item #1
    </div>
    <div>
      Item #2
    </div>
    <div>
      Item #3
    </div>
    <div>
      Item #4
    </div>
    <div>
      Item #5
    </div>
  </div>
</div>
`;

exports[`MasonryGrid should be rendered correctly with 3 columns 1`] = `
<div
  className="row"
>
  <div
    className="col-4"
  >
    <div>
      Item #1
    </div>
    <div>
      Item #4
    </div>
  </div>
  <div
    className="col-4"
  >
    <div>
      Item #2
    </div>
  </div>
  <div
    className="col-4"
  >
    <div>
      Item #3
    </div>
    <div>
      Item #5
    </div>
  </div>
</div>
`;

exports[`MasonryGrid should be rendered correctly with 4 columns 1`] = `
<div
  className="row"
>
  <div
    className="col-3"
  >
    <div>
      Item #1
    </div>
    <div>
      Item #5
    </div>
  </div>
  <div
    className="col-3"
  >
    <div>
      Item #2
    </div>
  </div>
  <div
    className="col-3"
  >
    <div>
      Item #3
    </div>
  </div>
  <div
    className="col-3"
  >
    <div>
      Item #4
    </div>
  </div>
</div>
`;
